File: serve.js
#!/usr/bin/env node
/**
 * Simple HTTP server for development
 * UPDATED COMMENTS: Node.js development server with ES6 modules support
 * SCALED FOR: Hot reload and proper MIME types
 */

const http = require('http');
const fs = require('fs');
const path = require('path');
const url = require('url');

// REUSED: MIME type mapping for web assets
const mimeTypes = {
  '.html': 'text/html',
  '.js': 'application/javascript',
  '.mjs': 'application/javascript',
  '.css': 'text/css',
  '.json': 'application/json',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.gif': 'image/gif',
  '.svg': 'image/svg+xml',
  '.ico': 'image/x-icon',
  '.woff': 'font/woff',
  '.woff2': 'font/woff2',
  '.ttf': 'font/ttf',
  '.otf': 'font/otf'
};

/**
 * Get MIME type for file extension
 * UPDATED COMMENTS: Enhanced MIME type detection with fallback
 */
function getMimeType(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  return mimeTypes[ext] || 'application/octet-stream';
}

/**
 * Serve static files with proper headers
 * SCALED FOR: Development-friendly caching and CORS
 */
function serveFile(res, filePath) {
  fs.readFile(filePath, (err, data) => {
    if (err) {
      res.writeHead(404, { 'Content-Type': 'text/plain' });
      res.end('File not found');
      return;
    }
    
    const mimeType = getMimeType(filePath);
    const headers = {
      'Content-Type': mimeType,
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
      'Access-Control-Allow-Headers': '*'
    };
    
    // CRITICAL: No cache for development
    if (mimeType === 'application/javascript' || mimeType === 'text/css') {
      headers['Cache-Control'] = 'no-cache, no-store, must-revalidate';
      headers['Pragma'] = 'no-cache';
      headers['Expires'] = '0';
    }
    
    res.writeHead(200, headers);
    res.end(data);
  });
}

/**
 * Main request handler
 * REUSED: Standard HTTP server request handling
 */
function requestHandler(req, res) {
  // UPDATED COMMENTS: Handle CORS preflight requests
  if (req.method === 'OPTIONS') {
    res.writeHead(200, {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
      'Access-Control-Allow-Headers': '*'
    });
    res.end();
    return;
  }
  
  const parsedUrl = url.parse(req.url);
  let pathname = parsedUrl.pathname;
  
  // SCALED FOR: Default file serving
  if (pathname === '/') {
    pathname = '/index.html';
  }
  
  const filePath = path.join(__dirname, pathname);
  
  // CRITICAL: Security check - prevent directory traversal
  if (!filePath.startsWith(__dirname)) {
    res.writeHead(403, { 'Content-Type': 'text/plain' });
    res.end('Forbidden');
    return;
  }
  
  // REUSED: File existence check
  fs.access(filePath, fs.constants.F_OK, (err) => {
    if (err) {
      res.writeHead(404, { 'Content-Type': 'text/plain' });
      res.end('File not found: ' + pathname);
      console.log('404:', pathname);
      return;
    }
    
    serveFile(res, filePath);
    console.log('200:', pathname);
  });
}

// ANCHOR: server_startup
const PORT = process.env.PORT || 8080;
const server = http.createServer(requestHandler);

server.listen(PORT, () => {
  console.log(`Server running at http://localhost:${PORT}/`);
  console.log(`Debug page: http://localhost:${PORT}/debug.html`);
  console.log('Press Ctrl+C to stop');
});

// UPDATED COMMENTS: Graceful shutdown
process.on('SIGINT', () => {
  console.log('\nServer stopped');
  process.exit(0);
});

File: serve.py
#!/usr/bin/env python3
"""
Simple HTTP server for development
UPDATED COMMENTS: Local development server with proper MIME types
SCALED FOR: CORS handling and ES6 modules support
"""

import http.server
import socketserver
import os
import mimetypes
from urllib.parse import urlparse

# REUSED: Custom MIME types for modern web development
mimetypes.add_type('application/javascript', '.js')
mimetypes.add_type('text/css', '.css')
mimetypes.add_type('application/json', '.json')
mimetypes.add_type('image/svg+xml', '.svg')

class CORSHTTPRequestHandler(http.server.SimpleHTTPRequestHandler):
    """
    HTTP Request Handler with CORS support
    UPDATED COMMENTS: Enables cross-origin requests for development
    CRITICAL: SPA routing - fallback to index.html for client-side routes
    """
    
    def do_GET(self):
        # CRITICAL: SPA routing fallback
        # UPDATED COMMENTS: If path doesn't exist and is not a file request, serve index.html
        parsed_path = urlparse(self.path)
        path = parsed_path.path
        
        # CRITICAL: Get basename and check for file extension
        basename = os.path.basename(path)
        has_extension = '.' in basename and not basename.startswith('.')
        
        # SCALED FOR: Check if file exists
        if path != '/' and not os.path.exists('.' + path):
            # CRITICAL: Only serve index.html for routes WITHOUT file extensions
            # UPDATED COMMENTS: If it has extension (.js, .css, .jpg, etc), let it 404
            if not has_extension:
                # REUSED: Serve index.html for SPA routes
                self.path = '/index.html'
        
        return super().do_GET()
    
    def end_headers(self):
        # CRITICAL: CORS headers for development
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', '*')
        
        # SCALED FOR: Caching control
        if self.path.endswith('.js'):
            self.send_header('Cache-Control', 'no-cache, no-store, must-revalidate')
        elif self.path.endswith('.css'):
            self.send_header('Cache-Control', 'no-cache, no-store, must-revalidate')
        elif self.path.endswith('.html') or self.path == '/':
            self.send_header('Cache-Control', 'no-cache, no-store, must-revalidate')
            self.send_header('Pragma', 'no-cache')
            self.send_header('Expires', '0')
        
        super().end_headers()
    
    def do_OPTIONS(self):
        # REUSED: OPTIONS method for CORS preflight
        self.send_response(200)
        self.end_headers()
    
    def guess_type(self, path):
        # UPDATED COMMENTS: Enhanced MIME type detection
        mimetype, encoding = mimetypes.guess_type(path)
        
        # CRITICAL: Ensure JavaScript modules are served correctly
        if path.endswith('.js'):
            return 'application/javascript'
        elif path.endswith('.mjs'):
            return 'application/javascript'
        
        return mimetype

def main():
    PORT = 8080
    
    # UPDATED COMMENTS: Change to project directory
    os.chdir(os.path.dirname(os.path.abspath(__file__)))
    
    with socketserver.TCPServer(("", PORT), CORSHTTPRequestHandler) as httpd:
        print(f"Server running at http://localhost:{PORT}/")
        print(f"Debug page: http://localhost:{PORT}/debug.html")
        print("Press Ctrl+C to stop")
        
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            print("\nServer stopped")

if __name__ == "__main__":
    main()